src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2858,98 +2855,423 @@ private static CommandBaseAst GetPreviousPipelineCommand(CommandAst commandAst)
+                }
+            }
 
-        public static bool AstAssignsToSameVariable(this VariableExpressionAst variableAst, Ast ast)
-        {
-            var parameterAst = ast as ParameterAst;
-            var variableAstVariablePath = variableAst.VariablePath;
-            if (parameterAst != null)
+            private bool AssignsToTargetVar(VariableExpressionAst foundVar)
             {
-                return variableAstVariablePath.IsUnscopedVariable &&
-                       parameterAst.Name.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase) &&
-                       parameterAst.Parent.Parent.Extent.EndOffset > variableAst.Extent.StartOffset;
+                if (!foundVar.VariablePath.UnqualifiedPath.EqualsOrdinalIgnoreCase(VariableTarget.VariablePath.UnqualifiedPath))
+                {
+                    return false;
+                }
+
+                int scopeIndex = foundVar.VariablePath.UserPath.IndexOf(':');
+                string scopeName = scopeIndex == -1 ? string.Empty : foundVar.VariablePath.UserPath.Remove(scopeIndex);
+                return AssignsToTargetScope(scopeName);
             }
 
-            if (ast is ForEachStatementAst foreachAst)
+            private bool AssignsToTargetVar(string userPath)
             {
-                return variableAstVariablePath.IsUnscopedVariable &&
-                       foreachAst.Variable.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase);
+                if (string.IsNullOrEmpty(userPath))
+                {
+                    return false;
+                }
+
+                string scopeName;
+                string varName;
+                int scopeIndex = userPath.IndexOf(':');
+                if (scopeIndex == -1)
+                {
+                    scopeName = string.Empty;
+                    varName = userPath;
+                }
+                else
+                {
+                    scopeName = userPath.Remove(scopeIndex);
+                    varName = userPath.Substring(scopeIndex + 1);
+                }
+
+                if (!varName.EqualsOrdinalIgnoreCase(VariableTarget.VariablePath.UnqualifiedPath))
+                {
+                    return false;
+                }
+
+                return AssignsToTargetScope(scopeName);
             }
 
-            if (ast is CommandAst commandAst)
+            private bool AssignsToTargetScope(string scopeName)
             {
-                string[] variableParameters = { "PV", "PipelineVariable", "OV", "OutVariable" };
-                StaticBindingResult bindingResult = StaticParameterBinder.BindCommand(commandAst, false, variableParameters);
+                if (LocalScopeOnly)
+                {
+                    return string.IsNullOrEmpty(scopeName) || scopeName.EqualsOrdinalIgnoreCase("Local") || scopeName.EqualsOrdinalIgnoreCase("Private");
```

Comment:
Do we really need to check for "Local" and "Private" if LocalScopeOnly is already true?

Reply:
Yes. `LocalScopeOnly` is for the variable we are trying to infer. These checks are for all the different variable assignments to see if they are related to the variable we are trying to infer or not.